### PR TITLE
I2C encoders

### PR DESCRIPTION
**Pull request by jaysonkelly**
This code supports Magnetic encoders which use the I2C communication protocol to communicate real-time position data to the mainboard. I believe I have finally got it correcting for skips only as opposed to correcting for every single tiny vibration it would identify as error. The encoder boards were initially developed by Aus3D and can be purchased from him or from Reliabuild3D along with the supporting hardware. The wiki is found here: http://wiki.aus3d.com.au/Magnetic_Encoder 

This is ready to be released to the community and the Marlin team will probably want to be made aware of the hardware should they choose to support it.

### CODE REVIEW

**Comment by teemuatlut** [comment]
Cool. I want one. Two.
A bit pricey though so I'll have to see if it comes down with time.

**Comment by bgort** [comment]
I'm impressed.  0.480um resolution?  That's pretty good....

**Comment by bgort** [comment]
@jaysonkelly Sounds like the chip he's been using has been discontinued (the AS5311), and though there's a replacement (NSE5310), it hasn't been tested?  Is that still the case?

Looks like these are out of stock.  Do you know if he's planning on updating the board design (if necessary), testing the NSE5310, and continuing production, etc. ?

Sorry to pepper you with questions, but pretty interesting.  Thinking about buying some and/or making them when I get time.

**Comment by chrissbarr** [comment]
Hi @bgort, Chris from Aus3D here. I actually got a bit mixed up when I was writing about the AS5311 and the NSE5310, it turned out that the '5311 is the newer model. I thought the '5311 had been discontinued, but it turned out it was just a supply shortage because they were a new product. 

A batch was made up and sold using the older '5310, but the design and firmware have been updated to the '5311 and I should have a batch ready to go in a few weeks. @jaysonkelly has been working with the updated '5311 modules for a while now, and ReliaBuild3D (Jayson and co) are using them on their printers.

Performance-wise I've not seen any significant difference between the two chips, it seems like the manufacturer just dropped I2C for a faster non-addressable serial interface. From the user's POV (or the printer's firmware) there's no difference between the '5310 and '5311 based modules, as an onboard micro handles the direct communication with the encoder.

@jaysonkelly has been doing some great work overhauling and building on the initial Marlin code I added. The I2C communication is a lot faster and more fault-resistant than it was, and I'm looking forward to playing with this code with the new encoder modules shortly.

It would be great to see this pulled into Marlin, though I know supporting uncommon hardware can sometimes raise questions of whether it's worth the work to integrate and maintain things. I think most of the changes are pretty non-invasive, but I wonder if it would make more sense for the encoder configuration section to be in the advanced config file? That way there's no visible change to the majority of people who aren't using this, and that might make it a bit more palatable.

@jaysonkelly I notice the address change and version check g-codes have been disabled - I'm out of the loop on this (or I'm forgetting something), is there a problem with those sections?

**Comment by emartinez167** [comment]
I agree that putting it into the advanced configuration file will make it easier to be adopted. Most uncommon stuff is there. I would submit the options as disabled, and update the RigidBot 2 sample configurations as needed.

Regards,

Ernesto.

**Comment by jaysonkelly** [comment]
@chrisbarr I have commented out everything that is not required most of the time to reduce the impact on dynamic memory. Somehow those functions use up quite a bit.

**Comment by chrisbarr** [comment]
👍

**Comment by bgort** [comment]
Hi @chrissbarr.  Thank you for getting back to me.  

> it turned out that the '5311 is the newer model. I thought the '5311 had been discontinued, but it turned out it was just a supply shortage because they were a new product.

Ah, good to hear!  That makes things easier...

> A batch was made up and sold using the older '5310, but the design and firmware have been updated to the '5311 and I should have a batch ready to go in a few weeks.

Great.  I'll check back in in a few weeks.

> From the user's POV (or the printer's firmware) there's no difference between the '5310 and '5311 based modules, as an onboard micro handles the direct communication with the encoder.

Makes sense.  I imagined you could likely deal with any differences on the encoder board's 328 and so present a consistent interface to the outside world.

> @jaysonkelly has been doing some great work overhauling and building on the initial Marlin code I added. The I2C communication is a lot faster and more fault-resistant than it was, and I'm looking forward to playing with this code with the new encoder modules shortly.

I've only skimmed through it, but I didn't see any major problems with anything.  Had a few comments, which I'll give below and in the code.

> It would be great to see this pulled into Marlin, though I know supporting uncommon hardware can sometimes raise questions of whether it's worth the work to integrate and maintain things. I think most of the changes are pretty non-invasive, but I wonder if it would make more sense for the encoder configuration section to be in the advanced config file? That way there's no visible change to the majority of people who aren't using this, and that might make it a bit more palatable.

I'm personally interested in making my printer as precise, robust, etc. as possible, and so this is of interest to me.  I know a lot of others are interested in the same things, so I see no problem with adding this to Marlin and can help maintain it, assuming there are no objections from others.  

A few questions:
- It looks like the resolution of the encoders is roughly half a micron, which is fantastic, but what are you seeing in terms of actual accuracy?  I haven't spent a lot of time with the code, but I saw mentions of error, error counting, error management schemes, etc.?
- Somewhat personal question:  I saw that magnetic strips are limited to 300mm in length.  I assume this isn't a technical limit, but perhaps just a supply issue?  Where are you at on getting longer ones?  I have a 312mm axis and I imagine some padding on either end is a good idea; would be nice to be able to make use of the whole thing.  I imagine others are in a similar position, so support for axes >300mm would be good.

My few thoughts on the PR itself:
- I do agree that this should likely go in the advanced configuration file, as there aren't many using it right now.  
- Rather than comment out the sections of the code that @jaysonkelly has mentioned take up a lot of memory and aren't necessary for normal operation (e.g. M864, M865), we could use a #define to conditionally compile them; maybe something like I2C_ENCODERS_CONFIGMODE (or something shorter)
- ... more in a pending review of the changes proposed in the PR


EDIT:  I don't think the CONFIGMODE idea is necessary now, after having reviewed the code; I suspect most of the problem is due to strings being held in memory instead of progmem.

File: Marlin/Configuration.h
Comment (bgort) [inline Marlin/Configuration.h]: Maybe more commentary about what this is, how it works both in theory and practically, that it's meant to be used with encoder boards using the '5311, details on their resolution, accuracy, etc.  Technical details would help, generally, I think.

File: Marlin/Configuration.h
Comment (bgort) [inline Marlin/Configuration.h]: Maybe I2C_POSITION_ENCODERS?  This way it's clearer what it is everywhere across the codebase.

File: Marlin/Configuration.h
Comment (bgort) [inline Marlin/Configuration.h]: Use the ENABLED() macro here, as in #if ENABLED(I2C_POSITION_ENCODERS)

File: Marlin/Marlin_main.cpp
Comment (bgort) [inline Marlin/Marlin_main.cpp]: Use SERIAL_ECHOPGM, SERIAL_ECHOLNPGM, SERIAL_ECHOPAIR, SERIAL_ECHOLNPAIR, etc. for all output, so we can rely on progmem.  I'll mark a few of the others, but these should all be changed.  This is a likely source of some of the memory usage @jaysonkelly is seeing.

File: Marlin/Marlin_main.cpp
Comment (bgort) [inline Marlin/Marlin_main.cpp]: Use progmem.

File: Marlin/i2cEncoder.cpp
Comment (bgort) [inline Marlin/i2cEncoder.cpp]: progmem...

File: Marlin/i2cEncoder.cpp
Comment (bgort) [inline Marlin/i2cEncoder.cpp]: progmem... 

(okay, I'll stop mentioning these, though there are quite a few more later on in the code..)

File: Marlin/i2cEncoder.cpp
Comment (bgort) [inline Marlin/i2cEncoder.cpp]: What's this (the ECM_PLANNER case) for?  Will it ever be used?

File: Marlin/i2cEncoder.cpp
Comment (bgort) [inline Marlin/i2cEncoder.cpp]: Hmm.  Are there problems with reliability, or is this just a remnant of early testing?

File: Marlin/Configuration.h
Comment (bgort) [inline Marlin/Configuration.h]: I see support for up to 5 encoders; should the 4th and 5th be added here?

File: Marlin/i2cEncoder.h
Comment (bgort) [inline Marlin/i2cEncoder.h:31]: Perhaps these should go into the advanced configuration along with the other defines...
(the things that are optional or likely to be altered, at least)

File: Marlin/i2cEncoder.h
Comment (bgort) [inline Marlin/i2cEncoder.h]: These can be probably a one line macro and can be added to macros.h; it doesn't look like it already exists.

Maybe call it COMBCOMP or SPACESHIP (PHP has an operator for this - <=> - which they call a spaceship operator).  Or something like that.

**Review by bgort — CHANGES_REQUESTED** [review]
Looking good, so far.  Mostly just minor things.

**Comment by bgort** [comment]
I'll probably have more feedback, etc., once I've tried the encoders, used the code/features, etc., but I don't see any reason not to incorporate sooner as long as whatever final version we end up with is conditionally compiled (obviously it is), tested, and working...

**Comment by jaysonkelly** [comment]
Awesome feedback! Thanks. I'll start working on all that as soon as I get the chance.

**Comment by bgort** [comment]
Great!

**Comment by jaysonkelly** [comment]
@bgort The magnetic strips actually come in continuous lengths from the manufacturer, so it really isn't a problem to make them any length needed. The folks at Laser Weld Design have about 80 encoder boards in stock and enough magnetic strip for several dual axis kits. We have tweaked Chris' design to replace the tiny connector with a through hole style to increase reliability and even though the bigger connector makes it a little more bulky, we have had very good results. If you are interested in helping with development I'm sure we can work something out to get you a kit. Go to Reliabuild3D.com and send us a message with the info on the contact page. While you are there you can check out the rest of the site and give us your thoughts.

File: Marlin/Configuration.h
Comment (jaysonkelly) [reply Marlin/Configuration.h]: Great idea. Probably include links to the wiki and the github for the encoder board.

File: Marlin/Marlin_main.cpp
Comment (jaysonkelly) [reply Marlin/Marlin_main.cpp]: I've never used those macros before. Are they all in macros.h?

**Comment by bgort** [comment]
I will check out the site.  Thanks.

File: Marlin/Marlin_main.cpp
Comment (bgort) [reply Marlin/Marlin_main.cpp]: They're in serial.h.

File: Marlin/Configuration.h
Comment (bgort) [reply Marlin/Configuration.h]: Yes, sounds good.

File: Marlin/i2cEncoder.cpp
Comment (jaysonkelly) [reply Marlin/i2cEncoder.cpp]: With the right setup there are no problems, but this was included in the early code to allow for and manage the possibility that the magnet looses position because it is out of range of the magnetic strip during operation. In my experience this added more issues that what it may have resolved. If the magnetic strip looses position during the print in the current code my thoughts are that the encoders should get disabled and the user should be notified that this is happening by lcd message and serial. @chrisbarr thoughts?

File: Marlin/i2cEncoder.cpp
Comment (jaysonkelly) [reply Marlin/i2cEncoder.cpp]: We have found that the babystepping method works very well and provides an immediate adjustment to the position. I think the other methods can be safely removed. @chrisbarr what do you think?

File: Marlin/Marlin_main.cpp
Comment (jaysonkelly) [reply Marlin/Marlin_main.cpp]: That makes a huge difference.

Comment (jaysonkelly) [reply Marlin/Marlin_main.cpp]: Seems like I still need to use SERIAL_ECHO and SERIAL_ECHOLN when sending variables. Is that correct?

Comment (bgort) [reply Marlin/Marlin_main.cpp]: Yes.  Variables are coming directly from SRAM, so you can't use the PGM/progmem functions, which pull from the flash.

File: Marlin/i2cEncoder.h
Comment (jaysonkelly) [reply Marlin/i2cEncoder.h]: You mean the sign functions? What would a single line macro look like for this?

**Comment by jaysonkelly** [comment]
@bgort Made a few tweaks.  What do you think?

File: Marlin/i2cEncoder.h
Comment (jaysonkelly) [reply Marlin/i2cEncoder.h:31]: I moved a few things, but the majority will never be messed with.

File: Marlin/i2cEncoder.h
Comment (bgort) [reply Marlin/i2cEncoder.h]: Apologies - I somehow thought these were a combined comparison functions when I looked at them the first time.  I see now that they're just returning the sign of the variable.  Sorry about that.

So you'd want something like this: 
`#define SIGN(a) ((a>0) - (a<0))`

Can just use that in place of the sgn() functions; no need to worry about types, etc.

If that produces a compiler warning anywhere (I don't think it should), you can also use a pair of ternaries, which should be fine (but maybe a little slower):
`#define SIGN(a) (a<0 ? -1 : (a>0 ? 1 : 0))`

**Comment by bgort** [comment]
I'll review this properly on Tuesday -- out of town until then.  Thanks for making the changes.

**Comment by bgort** [comment]
This is looking good.  Looks like the SIGN macro is left, and maybe a few other little tweaks?  

A rebase is needed now, too; sorry about that - I added something to the configuration that conflicts, though this one is an easy fix..

We're probably getting closer to the point we can merge this, assuming it all works after the changes?  Once it's included it'll be easier for others to tweak/improve/etc. ..

**Comment by jaysonkelly** [comment]
Sweet. Thanks for your help. I'll see what I can do about the macro soon.

**Comment by bgort** [comment]
I just fixed the conflict, but then noticed that the RCBugFix branch is targeted for the PR.  You'll need to do another PR targeting the bugfix-1.1.x branch, as I don't think there's any way to change a PR target...

**Comment by jaysonkelly** [comment]
@bgort Looks like the sign code is only used in one line that was actually commented out. It was a line left over from earlier code versions. Is there anything left to do before we merge it?

**Comment by jaysonkelly** [comment]
A PR to the 1.1.x bugfix has a few conflicts. I'm not sure how to resolve them as I've never done anything with the command line.  Thoughts?

**Comment by bgort** [comment]
Looking good!  

I'll see about pulling down your changes, applying them to the latest bugfix, and will then resolve the conflicts and create a new PR targeting bugfix.  Will have to be tomorrow or the next day, as I'm swamped today.

**Comment by jaysonkelly** [comment]
Awesome! I have your kit ready. Should ship this week.

**Comment by Roxy-3D** [comment]
> Awesome! I have your kit ready. Should ship this week.

If you ship two kits... I'll help @bgort 's brain resonant...   It would be super cool to not just assume we know the position of a stepper motor just because we told it to go somewhere.    I'm not even sure where this goes...  But as an example... Travel moves can happen at full speed (or over full speed) because if there is an out of position error... Well, we just fix it....    

This opens up a whole new horizon.

**Comment by jaysonkelly** [comment]
@Roxy-3d We would be happy to put a development kit together for you. Email us at 3dp@laserwelddesign.com 
Let us know how many encoder modules you want, lengths of magnetic strip and tell us about your cabling requirements and we will get you a quote with the same discounts we gave bgort.

**Comment by bgort** [comment]
> This opens up a whole new horizon.

It really does.  Potentially changes a -lot- of things..

**Comment by bgort** [comment]
You know, I imagine with this we could do away with 'steps per unit' and other motion-related settings.  If the magnetic strips are consistent and the encoders are as reliable, accurate, etc. as advertised, things like that could be calculated, or probably just ignored outright.  For regular printing, counting steps could be useless/unnecessary.  We could also calculate maximum speed, acceleration, etc. before steps are lost, and even look beyond that (jerk, snap, ..).  And as a next step, we could mount MEMS-based accelerometers on the axes and use their output to find the capabilities and limits of the steppers, drivers, etc., though that's maybe a bit overkill.  Would be neat, though.

With even basic capability, there'll be no more worrying about backlash or belt stretch or anything like that.  Prints could be nearly 100% dimensionally perfect and consistent.

Whole 'nother universe....

Anyway, I'm going to see about getting this moved over to the latest bugfix today.  More later...

**Comment by bgort** [comment]
Closing this as it's been replaced by #6946.

**Pull request closed without merging**
